src/infra/core/src/repos/metadata_chain_validators.rs

```diff
@@ -323,3 +341,462 @@ impl<'a> MetadataChainVisitor for ValidateOffsetsAreSequentialVisitor<'a> {
+
+                if e.is_empty() {
+                    return Err(
+                        AppendValidationError::no_op_event(e.clone(), "Event is empty").into(),
+                    );
+                }
+
+                Ok(Decision::NextOfType(state.next_block_flags))
+            }
+            State::ExecuteTransform(state) => {
+                let e = state.appended_execute_transform;
+
+                if e.is_empty() {
+                    return Err(
+                        AppendValidationError::no_op_event(e.clone(), "Event is empty").into(),
+                    );
+                }
+
+                Ok(Decision::NextOfType(state.next_block_flags))
+            }
+            State::SetPollingSource {
+                appended_set_polling_source: e,
+                appended_event,
+            } => {
+                // Queries must be normalized
+                if let Some(transform) = &e.preprocess {
+                    Self::validate_transform(appended_event, transform)?;
+                }
+
+                // Ensure no active push sources
+                Ok(Decision::NextOfType(Flag::ADD_PUSH_SOURCE))
+            }
+            State::AddPushSource {
+                appended_add_push_source: e,
+                appended_event,
+            } => {
+                // Ensure specifies the schema
+                if e.read.schema().is_none() {
+                    invalid_event!(
+                        (*e).clone(),
+                        "Push sources must specify the read schema explicitly",
+                    );
+                }
+
+                // Queries must be normalized
+                if let Some(transform) = &e.preprocess {
+                    Self::validate_transform(appended_event, transform)?;
+                }
+
+                Ok(Decision::NextOfType(Flag::SET_POLLING_SOURCE))
+            }
+            State::SetTransform {
+                appended_set_transform: e,
+                appended_event,
+            } => {
+                // Ensure has inputs
+                if e.inputs.is_empty() {
+                    invalid_event!((*e).clone(), "Transform must have at least one input");
+                }
+
+                // Ensure inputs are resolved to IDs and aliases are specified
+                for i in &e.inputs {
+                    if i.dataset_ref.id().is_none() || i.alias.is_none() {
+                        invalid_event!(
+                            (*e).clone(),
+                            "Transform inputs must be resolved to dataset IDs and specify aliases"
+                        );
+                    }
+                }
+
+                // Queries must be normalized
+                Self::validate_transform(appended_event, &e.transform)?;
+
+                Ok(Decision::Stop)
+            }
+            State::Stopped => Ok(Decision::Stop),
+        }
+    }
+
+    fn visit(&mut self, (_, block): HashedMetadataBlockRef) -> Result<Decision, Self::Error> {
```

Comment:
Yikes! This visitor is a no-go. My eyes are bleeding.

The entire problem of `validate_append_event_logical_structure` was that there was a ton of rules it needed to enforce, but we couldn't separate it into multiple rules because there was no mechanism to run them in a **single pass over metadata**.

In other words the complexity of `validate_append_event_logical_structure` was exactly why we needed the visitor mechanism.

Reply:
Roughly speaking, we could take every error message in `validate_append_event_logical_structure` and turn the conditions under which it is generated into a rule with its own visitor.

For example extracting visitors like:
- `ValidateEventIsNotEmptyOrNoOp`
- `ValidateDataSchemaIsSetBeforeAddingData`
- `ValidateCheckpointSequencing`
- `ValidateWatermarkSequencing`
- `ValidateTransformInputOrderIsStable`
- ...

Reply:
The visitor was shredded to pieces